In [6]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import train_test_split
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
import scipy as sp
import statsmodels.formula.api as sm
from sklearn.inspection import permutation_importance
from itertools import product
TAMANHO_TESTE = 0.2
SEMENTE_ALEATORIA = 1
DATASET_NAME = "Minerals_Database3.csv"
FEATURES = ["Specific Gravity", "Mohs Hardness", "Dispersion", 'Molar Mass', 'Molar Volume', 'Calculated Density']
TARGET = ["Refractive Index"]

df = pd.read_csv("Minerals_Database3.csv")
indices = df.index
indices_treino, indices_teste = train_test_split(
    indices, test_size=TAMANHO_TESTE, random_state=SEMENTE_ALEATORIA
)

df_treino = df.loc[indices_treino]
df_teste = df.loc[indices_teste]

X_treino = df_treino.reindex(FEATURES, axis=1).values
y_treino = df_treino.reindex(TARGET, axis=1).values.ravel()
X_teste = df_teste.reindex(FEATURES, axis=1).values
y_teste = df_teste.reindex(TARGET, axis=1).values.ravel()

In [10]:
from itertools import product
from sklearn.metrics import r2_score



NUM_ARVORES = [10, 20, 30, 40, 50]
NUM_FOLHAS = [10, 20, 30, 40, 50]
NUM_PROFUNDIDADE = [10, 20, 30]
NUM_FOLDS = 5
NUM_CPU_CORES = 4



for n_folhas, n_profundidade, n_arvores in product(
    NUM_FOLHAS, NUM_PROFUNDIDADE, NUM_ARVORES
):
    modelo_rf = RandomForestRegressor(
        n_estimators=n_arvores,
        max_leaf_nodes=n_folhas,
        max_depth=n_profundidade,
        random_state=SEMENTE_ALEATORIA,
        n_jobs= NUM_CPU_CORES,
    )
    
    scores = cross_val_score(
    modelo_rf,
    X_treino,
    y_treino.ravel(),
    cv=NUM_FOLDS,
    )

    print("Aplicando a métrica de erro R² para os seguintes hiperparâmetros: \n"
        f"{n_arvores} número de árvores,\n"
        f"{n_folhas} número de folhas e\n"
        f"{n_profundidade} número de profundidade\n"
        f"temos que:\n"
        f"Os scores foram de: {scores}\n"
        f"A média dos scores é de: {scores.mean()}"
        f"\n"
    )
    

#for n_folhas, n_profundidade, n_arvores in product(
    #NUM_FOLHAS, NUM_PROFUNDIDADE, NUM_ARVORES
#):
    
    #modelo_rf = RandomForestRegressor(
        #n_estimators=NUM_ARVORES,
        #random_state=SEMENTE_ALEATORIA,
        #n_jobs=NUM_CPU_CORES,
#)
    #scoresWithRMSE = cross_val_score(
    #modelo_rf,
    #X_treino,
    #y_treino,
    #cv=NUM_FOLDS,
    #scoring="neg_root_mean_squared_error",
#)


#print(
        #f"Aplicando a métrica de erro R² para os seguintes hiperparâmetros: \n"
        #f"{n_arvores} número de árvores,\n"
        #f"{n_folhas} número de folhas e\n"
        #f"{n_profundidade} número de profundidade\n"
        #f"temos que:\n"
        #f"Os scores foram de: {scoresWithRMSE}\n"
        #f"A média dos scores é de: {scoresWithRMSE.mean()}"
        #f"\n"
    #)

Aplicando a métrica de erro R² para os seguintes hiperparâmetros: 
10 número de árvores,
10 número de folhas e
10 número de profundidade
temos que:
Os scores foram de: [0.72382344 0.78593438 0.84310089 0.84234495 0.82629941]
A média dos scores é de: 0.8043006147643779

Aplicando a métrica de erro R² para os seguintes hiperparâmetros: 
20 número de árvores,
10 número de folhas e
10 número de profundidade
temos que:
Os scores foram de: [0.72002183 0.79465983 0.845084   0.84155001 0.83449645]
A média dos scores é de: 0.8071624228719463

Aplicando a métrica de erro R² para os seguintes hiperparâmetros: 
30 número de árvores,
10 número de folhas e
10 número de profundidade
temos que:
Os scores foram de: [0.71803752 0.79532381 0.84616345 0.84397495 0.83764647]
A média dos scores é de: 0.8082292399817327

Aplicando a métrica de erro R² para os seguintes hiperparâmetros: 
40 número de árvores,
10 número de folhas e
10 número de profundidade
temos que:
Os scores foram de: [0.71897649 0.79768389

## Há overfitting?

#### Nota-se que as médias ficaram de R^2 ficaram entre 0.75 e 0.81. Em primeiro sinal, isso parece bom, mas provavelmente está ocorrendo overfitting pois, no bloco II, o melhor modelo foi o de Baseline, que não é particularmente sofisticado. Provavelmente isso ocorre pois existe algo errado com a proporção de dados de teste e treino. Em geral, é possível que haja poucos dados de treino.

## Você recomendaria seu modelo?

#### Depende (como diriam os engenheiros de material). Para aplicação na área de estudo em que o dataset foi pronveniente? Não, uma vez que nós tratamos o dataset ao ponto de ele ficar pequeno e ideal. Essencialmente, cortamos a maior parte do dataset, o que simplificou e muito os modelos. Além disso, o overfitting é claro, então antes de tudo corrigir esse problema se faz necessário. Já para aplicações didáticas ele é perfeito, simplesmente por ser pequeno e ideal, assim quem se interessar pode se focar em como os modelos, algoritmos e métodos funcionam, sem se preocupar explicitamente com o tratamento do dataset. Naturalmente, é também possível utilizar o dataset bruto, de modo que a pessoa poderá aprender a tratá-lo.